vmaas_sync/system_culling.go

```diff
@@ -4,23 +4,41 @@ import (
 	"app/base"
 	"app/base/database"
 	"app/base/utils"
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
 	"time"
 )
 
+// Need to run code within a function, because defer can't be used in loops
+func withTx(do func(db *gorm.DB) error) error {
+	tx := database.Db.BeginTx(base.Context, nil)
+	defer tx.RollbackUnlessCommitted()
+	if err := do(tx); err != nil {
+		return err
+	}
+	return errors.Wrap(tx.Commit().Error, "Commit")
```

Comment:
that's a little confusing, we rely on errors.Wrap returns Nil on Nil input argument.